File: legacy/lethe/reranker.py
"""Cross-encoder reranker with adaptive depth."""
from __future__ import annotations

from typing import Any

import numpy as np


class Reranker:
    """Wraps a cross-encoder model with adaptive depth logic.

    Shallow pass: score k_fetch candidates.
    If max score < threshold, trigger deep pass: score k_deep candidates.
    """

    def __init__(
        self,
        cross_encoder: Any,
        confidence_threshold: float = 4.0,
    ) -> None:
        self.cross_encoder = cross_encoder
        self.confidence_threshold = confidence_threshold
        self._loaded = cross_encoder is not None

    def rerank(
        self,
        query_text: str,
        candidates: list[tuple[str, str]],  # (entry_id, content)
    ) -> list[tuple[str, float]]:
        """Score candidates with cross-encoder, return sorted (id, score) pairs."""
        if not self._loaded or not candidates:
            return [(eid, 0.0) for eid, _ in candidates]
        pairs = [(query_text, content) for _, content in candidates]
        scores = self.cross_encoder.predict(pairs)
        scored = [(eid, float(s)) for (eid, _), s in zip(candidates, scores)]
        scored.sort(key=lambda x: x[1], reverse=True)
        return scored

    def needs_deep_search(self, scores: list[float]) -> bool:
        """Returns True if the shallow pass didn't find anything confident."""
        if not scores:
            return True
        return max(scores) < self.confidence_threshold

File: crates/lethe-core/src/reranker.rs
//! Cross-encoder reranking + adaptive deep-pass — port of
//! `legacy/lethe/reranker.py`.
//!
//! `needs_deep_search` mirrors the Python check: trigger the deep
//! candidate pass when the top cross-encoder score from the shallow
//! batch is below `confidence_threshold`. Default 4.0 — picked in
//! checkpoint 6 and unchanged since.

use std::sync::Arc;

use crate::encoders::CrossEncoder;
use crate::Error;

/// Rerank shallow candidates, with the deep-pass decision baked in.
#[derive(Clone)]
pub struct Reranker {
    cross_encoder: Option<Arc<CrossEncoder>>,
    pub confidence_threshold: f32,
}

impl std::fmt::Debug for Reranker {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Reranker")
            .field("has_cross_encoder", &self.cross_encoder.is_some())
            .field("confidence_threshold", &self.confidence_threshold)
            .finish()
    }
}

impl Reranker {
    #[must_use]
    pub fn new(cross_encoder: Option<Arc<CrossEncoder>>, confidence_threshold: f32) -> Self {
        Self {
            cross_encoder,
            confidence_threshold,
        }
    }

    /// Score `(query, content)` pairs. Returns `(eid, score)` sorted
    /// descending by score. When no cross-encoder is configured, the
    /// candidates are returned in input order with score `0.0` for
    /// every entry (this is the test-only path; callers must not
    /// treat the score as meaningful in that case).
    pub fn rerank(
        &self,
        query: &str,
        candidates: &[(&str, &str)],
    ) -> Result<Vec<(String, f32)>, Error> {
        if candidates.is_empty() {
            return Ok(Vec::new());
        }
        let Some(xenc) = &self.cross_encoder else {
            return Ok(candidates
                .iter()
                .map(|(eid, _)| ((*eid).to_owned(), 0.0))
                .collect());
        };
        let pairs: Vec<(&str, &str)> = candidates.iter().map(|(_, c)| (query, *c)).collect();
        let scores = xenc.predict(&pairs)?;
        let mut scored: Vec<(String, f32)> = candidates
            .iter()
            .zip(scores)
            .map(|((eid, _), s)| ((*eid).to_owned(), s))
            .collect();
        scored.sort_by(|(a_id, a), (b_id, b)| {
            b.partial_cmp(a)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| a_id.cmp(b_id))
        });
        Ok(scored)
    }

    /// Adaptive deep-pass trigger: top score below threshold (or no
    /// scores at all). Mirrors `Reranker.needs_deep_search`.
    #[must_use]
    pub fn needs_deep_search(&self, scores: &[f32]) -> bool {
        let Some(top) = scores.first() else {
            return true;
        };
        *top < self.confidence_threshold
    }
}
